keps/sig-node/4569-cgroup-v1-maintenance-mode/README.md

```diff
@@ -0,0 +1,641 @@
+
+4. **Migration Support**: Provide clear and practical migration guidance for users using cgroup v1, facilitating a smoother transition to cgroup v2.
+
+5. **Enhancing cgroup v2 Support**: Address all known pending bugs in Kubernetes’ cgroup v2 support to ensure it reaches a level of reliability and functionality that encourages users to transition from cgroup v1.
+
+### Non-Goals
+
+Removing cgroup v1 support immediately.
+
+## Proposal
+
+The proposal outlines a plan to move cgroup v1 support in Kubernetes into maintenance mode, encouraging the community and users to transition to cgroup v2.
+
+### Risks and Mitigations
+
+The primary risk involves potential disruptions for users who migrate to cgroup v2 with incompatible workloads.
+
+Users depending on the following technologies will need to ensure they are using the specified versions or later, which support cgroup v2:
+
+- OpenJDK / HotSpot: jdk8u372, 11.0.16, 15 and later
+- NodeJs 20.3.0 or later
+- IBM Semeru Runtimes: jdk8u345-b01, 11.0.16.0, 17.0.4.0, 18.0.2.0 and later
+- IBM SDK Java Technology Edition Version (IBM Java): 8.0.7.15 and later
+- If users run any third-party monitoring and security agents that depend on the cgroup file system, they need to update the agents to a version that supports cgroup v2.
+
+Mitigations include providing clear documentation on the migration process, offering community support for common issues encountered during migration, and keeping the cgroup v1 support in the maintenance mode for allowing users additional time to switch to cgroup v2 without any major disruptions.
+
+
+## Design Details
+
+This enhancement outlines the steps required to transition existing cgroup v1 support into maintenance mode, and as such, no new feature gates are proposed in this document.
+
+### Introduction of cgroup version metric
+
+A new metric, `kubelet_cgroup_version`, is proposed. This metric will report values `1` or `2`, indicating whether the host is utilizing cgroup version `1` or `2`, respectively. The kubelet will assess the host's cgroup version at startup and emit this metric accordingly.
+
+The introduction of this metric aims to streamline the process for cluster administrators in determining the cgroup version deployed across their hosts. This metric removes the need for manual node inspection, providing a clear insight into the cgroup version each node operates on.
+
+### Implementing a warning log and an event for cgroup v1 usage
+
+Starting from 1.31, during kubelet startup if the host is running on cgroup v1, kubelet will log a warning message like,
+
+```golang
+klog.Warning("cgroup v1 support has been transitioned into maintenance mode, please plan for the migration towards cgroup v2")
+```
+and also emit a corresponding event,
+```golang
+eventRecorder.Event(pod, v1.EventTypeWarning, "CgroupV1", fmt.Sprint("cgroup v1 support has been transitioned into maintenance mode"))
+```
+
+#### Code modifications for default cgroup assumptions
```

Comment:
I know the feature gate has been dropped from the proposal, but I was just wondering if having a feature gate will make the code separation easier. It may increase the code readability for feature development, and can also prepare us for the next steps.

Reply:
https://github.com/kubernetes/enhancements/pull/4572#discussion_r1559911931 is how I'd approach that, with a feature gate that starts off _on_.

After at least one more minor release, things change: if you want to use cgroup v1 on a node, you manually need to flip the feature gate to on, because we've changed the default to _off_.

_time passes_

We drop the cgroup v1 support. Maybe. Or we leave it in place. I don't have an opinion on how that part plays out.